system/threads.cpp

```diff
@@ -115,44 +166,155 @@ long __syscall_futex(uint32_t* uaddr, int futex_op, ...)
 				if (*uaddr != val)
 				{
 					futexSpinLock.unlock();
-					return EAGAIN;
+					return -EAGAIN;
 				}
 				mainThreadWaitAddress.store(uaddr);
 				if (*uaddr != val)
 				{
 					mainThreadWaitAddress.store(0);
 					futexSpinLock.unlock();
-					return EAGAIN;
+					return -EAGAIN;
 				}
 				futexSpinLock.unlock();
-				int64_t startTime = 0;
-				if (timeout != -1)
+				addNodeToFutexList(&futexNode);
+				while (futexNode.sleepAddress != nullptr)
```

Comment:
sleepAddress (and next) should probably be atomic, and we should use load() and store() to access them

Reply:
With the new approach discussed during the sync-in, where every thread has a private address they sleep on, and a public address which is the one from userspace, we do not need the re-sleep mechanism anymore. Threads are only woken when they are truly woken, otherwise their public address will be changed but they will still just sleep on their private address.
I guess the variable we store the public address in, can also be the private address they sleep on?